# PROYECTO FINAL – NOTEBOOK ANALÍTICO CON SQL Y PYTHON
**Estudiante:** Oscar Gianfranco Navarro Navarro

**Módulo:** Introducción a SQL y Python para Análisis de Datos

**Título del proyecto:**
Análisis Exploratorio de Ventas y Rentabilidad de una Cadena de Supermercados

**Fuente del dataset:** Kaggle - Samples Superstore

**Descripción del proyecto:**

El propósito de este proyecto es examinar la información de ventas para descubrir patrones de compra, comprender el comportamiento de los clientes, evaluar el rendimiento de las diferentes regiones e identificar los factores que influyen en la rentabilidad del negocio.

**Objetivo:**

Realizar un análisis exploratorio del conjunto de datos de ventas de una tienda para identificar patrones de ventas, ganancias, descuentos, categorías de productos y comportamiento de las regiones mediante estadísticas descriptivas, visualizaciones interactivas y consultas SQL utilizando DuckDB.


## Carga del Dataset

Importamos las librerias

In [271]:
import pandas as pd
import plotly.express as px
import numpy as np

Cargamos el archivo CSV en un DataFrame de pandas

In [272]:
df = pd.read_csv("/content/Sample - Superstore.csv", encoding="latin1")

Visualizamos toda la información del DataFrame

In [273]:
df

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.2480,3,0.20,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.9600,2,0.00,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.5760,2,0.20,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.6000,4,0.00,13.3200


# Exploración Inicial
Revisamos información general del DataFrame (columnas, tipos de datos, valores no nulos)

Resumen detallado de la estructura del DataFrame

In [274]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

Tamaño del DataFrame
- 9994 filas → hay 9.994 registros de ventas.
- 21 columnas → existen 21 variables o características del dataset.

In [275]:
df.shape

(9994, 21)

Se visualiza las primeras 5 filas del DataFrame

In [276]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


Se visualiza las últimas 5 filas del DataFrame

In [277]:
df.tail()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.248,3,0.2,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.960,2,0.0,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.576,2,0.2,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.600,4,0.0,13.3200
9993,9994,CA-2017-119914,5/4/2017,5/9/2017,Second Class,CC-12220,Chris Cortes,Consumer,United States,Westminster,...,92683,West,OFF-AP-10002684,Office Supplies,Appliances,"Acco 7-Outlet Masterpiece Power Center, Wihtou...",243.160,2,0.0,72.9480


Se visualiza los nombres de todas las columnas (variables) que contiene el DataFrame.

In [278]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

# Análisis de calidad de datos
Identificar la cantidad exacta de valores nulos (vacíos) que existen en cada columna

In [279]:
df.isnull().sum()

,0
Row ID,0
Order ID,0
Order Date,0
Ship Date,0
Ship Mode,0
Customer ID,0
Customer Name,0
Segment,0
Country,0
City,0


identificar si existen valores nulos o ausentes en las columnas

In [280]:
df.isna().any()

,0
Row ID,False
Order ID,False
Order Date,False
Ship Date,False
Ship Mode,False
Customer ID,False
Customer Name,False
Segment,False
Country,False
City,False


Verificar si hay filas completamente duplicadas

In [281]:
df.duplicated().sum()

np.int64(0)

Comprobar si existe al menos un valor nulo

In [282]:
df.isnull().any().any()

np.False_

Permite identificar y contar cuántos valores están como cadenas de texto vacías ("") dentro de cada columna

In [283]:
(df == "").sum()

,0
Row ID,0
Order ID,0
Order Date,0
Ship Date,0
Ship Mode,0
Customer ID,0
Customer Name,0
Segment,0
Country,0
City,0


Permite conocer el tipo de dato de cada columna
- Las variables numéricas (Sales, Profit, Discount, Quantity) estén listas para cálculos.
- Las variables categóricas (Category, Region, Segment) puedan utilizarse en agrupaciones y gráficos.
- Las fechas puedan convertirse correctamente para análisis temporal.

In [284]:
df.dtypes

,0
Row ID,int64
Order ID,object
Order Date,object
Ship Date,object
Ship Mode,object
Customer ID,object
Customer Name,object
Segment,object
Country,object
City,object


#Estadísticas descriptivas


Estadísticas descriptivas para variables numéricas

In [285]:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


Estadísticas descriptivas para variables categóricas

In [286]:
df.describe(include="object")

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Region,Product ID,Category,Sub-Category,Product Name
count,9994,9994,9994,9994,9994,9994,9994,9994,9994,9994,9994,9994,9994,9994,9994
unique,5009,1237,1334,4,793,793,3,1,531,49,4,1862,3,17,1850
top,CA-2017-100111,9/5/2016,12/16/2015,Standard Class,WB-21850,William Brown,Consumer,United States,New York City,California,West,OFF-PA-10001970,Office Supplies,Binders,Staple envelope
freq,14,38,35,5968,37,37,5191,9994,915,2001,3203,19,6026,1523,48


Resumen estadístico completo del DataFrame, incluyendo tanto las variables numéricas como las variables categóricas

In [287]:
df.describe(include='all')

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
count,9994.000000,9994,9994,9994,9994,9994,9994,9994,9994,9994,...,9994.000000,9994,9994,9994,9994,9994,9994.000000,9994.000000,9994.000000,9994.000000
unique,NaN,5009,1237,1334,4,793,793,3,1,531,...,NaN,4,1862,3,17,1850,NaN,NaN,NaN,NaN
top,NaN,CA-2017-100111,9/5/2016,12/16/2015,Standard Class,WB-21850,William Brown,Consumer,United States,New York City,...,NaN,West,OFF-PA-10001970,Office Supplies,Binders,Staple envelope,NaN,NaN,NaN,NaN
freq,NaN,14,38,35,5968,37,37,5191,9994,915,...,NaN,3203,19,6026,1523,48,NaN,NaN,NaN,NaN
mean,4997.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,55190.379428,NaN,NaN,NaN,NaN,NaN,229.858001,3.789574,0.156203,28.656896
std,2885.163629,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,32063.693350,NaN,NaN,NaN,NaN,NaN,623.245101,2.225110,0.206452,234.260108
min,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,1040.000000,NaN,NaN,NaN,NaN,NaN,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,23223.000000,NaN,NaN,NaN,NaN,NaN,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,56430.500000,NaN,NaN,NaN,NaN,NaN,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,90008.000000,NaN,NaN,NaN,NaN,NaN,209.940000,5.000000,0.200000,29.364000


Limpiar los nombres de las columnas

In [288]:
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
    .str.replace("-", "_")
)


Convertir columnas de fecha

In [289]:
df["order_date"] = pd.to_datetime(df["order_date"])
df["ship_date"] = pd.to_datetime(df["ship_date"])

df[["order_date", "ship_date"]].head()

,order_date,ship_date
0,2016-11-08,2016-11-11
1,2016-11-08,2016-11-11
2,2016-06-12,2016-06-16
3,2015-10-11,2015-10-18
4,2015-10-11,2015-10-18


Crear columnas de fechas para el análisis

In [290]:
df["order_year"] = df["order_date"].dt.year
df["order_month"] = df["order_date"].dt.month_name()
df["order_month_number"] = df["order_date"].dt.month
df["order_quarter"] = df["order_date"].dt.quarter
df["shipping_days"] = (df["ship_date"] - df["order_date"]).dt.days

Crear columna de Margen Profit

In [291]:
df["profit_margin"] = (df["profit"] / df["sales"]) * 100
df["profit_margin"] = df["profit_margin"].round(2)

df[["sales", "profit", "profit_margin"]].head()

,sales,profit,profit_margin
0,261.9600,41.9136,16.00
1,731.9400,219.5820,30.00
2,14.6200,6.8714,47.00
3,957.5775,-383.0310,-40.00
4,22.3680,2.5164,11.25


Comprobación posibles inconsistecias

In [292]:
df[["sales", "quantity", "discount", "profit", "shipping_days"]].describe()

,sales,quantity,discount,profit,shipping_days
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,229.858001,3.789574,0.156203,28.656896,3.958175
std,623.245101,2.225110,0.206452,234.260108,1.747567
min,0.444000,1.000000,0.000000,-6599.978000,0.000000
25%,17.280000,2.000000,0.000000,1.728750,3.000000
50%,54.490000,3.000000,0.200000,8.666500,4.000000
75%,209.940000,5.000000,0.200000,29.364000,5.000000
max,22638.480000,14.000000,0.800000,8399.976000,7.000000


In [293]:
df[df["sales"] <= 0]

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,sales,quantity,discount,profit,order_year,order_month,order_month_number,order_quarter,shipping_days,profit_margin


In [294]:
df[df["quantity"] <= 0]

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,sales,quantity,discount,profit,order_year,order_month,order_month_number,order_quarter,shipping_days,profit_margin


In [295]:
df[df["discount"] < 0]

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,sales,quantity,discount,profit,order_year,order_month,order_month_number,order_quarter,shipping_days,profit_margin


In [296]:
df[df["profit"] < 0]

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,sales,quantity,discount,profit,order_year,order_month,order_month_number,order_quarter,shipping_days,profit_margin
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,957.5775,5,0.45,-383.0310,2015,October,10,4,7,-40.00
14,15,US-2015-118983,2015-11-22,2015-11-26,Standard Class,HP-14815,Harold Pawlan,Home Office,United States,Fort Worth,...,68.8100,5,0.80,-123.8580,2015,November,11,4,4,-180.00
15,16,US-2015-118983,2015-11-22,2015-11-26,Standard Class,HP-14815,Harold Pawlan,Home Office,United States,Fort Worth,...,2.5440,3,0.80,-3.8160,2015,November,11,4,4,-150.00
23,24,US-2017-156909,2017-07-16,2017-07-18,Second Class,SF-20065,Sandra Flanagan,Consumer,United States,Philadelphia,...,71.3720,2,0.30,-1.0196,2017,July,7,3,2,-1.43
27,28,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,...,3083.4300,7,0.50,-1665.0522,2015,September,9,3,4,-54.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9920,9921,CA-2016-149272,2016-03-15,2016-03-19,Standard Class,MY-18295,Muhammed Yedwab,Corporate,United States,Bryan,...,22.3860,7,0.80,-35.8176,2016,March,3,1,4,-160.00
9921,9922,CA-2014-111360,2014-11-24,2014-11-30,Standard Class,AT-10435,Alyssa Tate,Home Office,United States,Akron,...,5.7420,3,0.70,-4.5936,2014,November,11,4,6,-80.00
9931,9932,CA-2015-104948,2015-11-13,2015-11-17,Standard Class,KH-16510,Keith Herrera,Consumer,United States,San Bernardino,...,683.3320,4,0.15,-40.1960,2015,November,11,4,4,-5.88
9937,9938,CA-2016-164889,2016-06-03,2016-06-06,Second Class,CP-12340,Christine Phan,Corporate,United States,Los Angeles,...,71.0880,2,0.20,-1.7772,2016,June,6,2,3,-2.50


# Visualizaciones con Plotly

**Gráfico 1 - Margen de Ganancia por Año**



El gráfico muestra la evolución del margen de ganancia promedio por año.

In [297]:
yearly_profit_margin = (
    df.groupby("order_year")["profit_margin"]
    .mean()
    .round(2)
    .reset_index()
)

fig = px.line(
    yearly_profit_margin,
    x="order_year",
    y="profit_margin",
    title="Profit Margin by Year",
    markers=True,
    text="profit_margin"
)

fig.update_traces(
    line_color="red",
    texttemplate="%{text:.2f}%",
    textposition="top center"
)

fig.update_layout(
    xaxis_title="Year",
    yaxis_title="Profit Margin (%)",
    template="plotly_white"
)

fig.show()

**Gráfico 2 - Ventas por Categoría**

Este gráfico interpreta que la categoría Technology es la que registra el mayor volumen de ventas, seguida por Furniture y Office Supplies. Esto indica que los productos tecnológicos representan la principal fuente de ingresos dentro del conjunto de datos analizado.

In [298]:
category_sales = (
    df.groupby("category")["sales"]
    .sum()
    .sort_values(ascending=False)
    .reset_index()
)

fig = px.bar(
    category_sales,
    x="category",
    y="sales",
    title="Sales by Category",
    color="category",
    text_auto=".2s"
)

fig.update_layout(
    xaxis_title="Category",
    yaxis_title="Total Sales",
    template="plotly_white"
)

fig.show()

**Gráfico 3 - Tendencia de ventas Mensuales**

Este gráfico representa la tendencia mensual de las ventas a lo largo del período analizado, permitiendo:
- Identificar meses con mayores o menores ventas.
- Detectar tendencias de crecimiento o disminución.
- Observar posibles patrones estacionales en el comportamiento de las ventas.

In [299]:
monthly_sales_chart["month_label"] = (
    monthly_sales_chart["order_month"].str[:3]
    + "-"
    + monthly_sales_chart["order_year"].astype(str)
)

fig = px.line(
    monthly_sales_chart,
    x="month_label",
    y="sales",
    title="Monthly Sales Trend",
    markers=True
)

fig.update_layout(
    xaxis_title="Month",
    yaxis_title="Sales",
    xaxis_tickangle=-45,
    template="plotly_white"
)

fig.show()

**Gráfico 4 - Los 5 principales estados con más ventas**

Este gráfico permite identificar:

- Los estados con mayor contribución al volumen de ventas.
- La concentración geográfica del negocio.
- Regiones con mayor importancia comercial.

In [300]:
top_states = (
    df.groupby("state")["sales"]
    .sum()
    .sort_values(ascending=False)
    .head(5)
    .reset_index()
)

fig = px.bar(
    top_states,
    x="state",
    y="sales",
    title="Top 5 States by Sales",
    color="sales",
    text="sales"
)

fig.show()

**Gráfico 5 - Distribución de segmentos de clientes**

Este gráfico permite identificar qué tipo de clientes tiene mayor participación en los ingresos del negocio.

In [301]:
segment_sales = (
    df.groupby("segment")["sales"]
    .sum()
    .reset_index()
)

fig = px.pie(
    segment_sales,
    values="sales",
    names="segment",
    title="Sales Distribution by Customer Segment",
    color_discrete_sequence=["yellow", "red", "blue"],
    hole=0.
)

fig.update_traces(
    textinfo="percent+label",
    textposition="inside"
)

fig.show()

# Consultas SQL con DuckDB

Importar la librería DuckDB

In [302]:
import duckdb

Crear una base de datos en memoria

In [303]:
con = duckdb.connect(database=":memory")

Registra el DataFrame df como una tabla virtual llamada store.

A partir de ese momento puedes escribir consultas SQL como si store fuera una tabla de una base de datos.

In [304]:
con.register("store",df)

Se puede dividir en dos partes:

- con.execute(): ejecuta la consulta SQL sobre la tabla store.
- fetch_df(): devuelve el resultado como un DataFrame de Pandas.

In [305]:
con.execute("SELECT * FROM store").fetch_df()

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,sales,quantity,discount,profit,order_year,order_month,order_month_number,order_quarter,shipping_days,profit_margin
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,261.9600,2,0.00,41.9136,2016,November,11,4,3,16.00
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,731.9400,3,0.00,219.5820,2016,November,11,4,3,30.00
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,14.6200,2,0.00,6.8714,2016,June,6,2,4,47.00
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,957.5775,5,0.45,-383.0310,2015,October,10,4,7,-40.00
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,22.3680,2,0.20,2.5164,2015,October,10,4,7,11.25
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,9990,CA-2014-110422,2014-01-21,2014-01-23,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,25.2480,3,0.20,4.1028,2014,January,1,1,2,16.25
9990,9991,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,91.9600,2,0.00,15.6332,2017,February,2,1,5,17.00
9991,9992,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,258.5760,2,0.20,19.3932,2017,February,2,1,5,7.50
9992,9993,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,29.6000,4,0.00,13.3200,2017,February,2,1,5,45.00


**1. Inspección de datos (SELECT + LIMIT)**

Permite visualizar una muestra de los 10 registros del dataset.

  ***Objetivo:**
Inspeccionar la estructura del dataset y verificar que los datos se hayan cargado correctamente.*

In [306]:
con.execute("""
SELECT *
FROM store
LIMIT 10
""").fetch_df()

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,sales,quantity,discount,profit,order_year,order_month,order_month_number,order_quarter,shipping_days,profit_margin
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,261.9600,2,0.00,41.9136,2016,November,11,4,3,16.00
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,731.9400,3,0.00,219.5820,2016,November,11,4,3,30.00
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,14.6200,2,0.00,6.8714,2016,June,6,2,4,47.00
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,957.5775,5,0.45,-383.0310,2015,October,10,4,7,-40.00
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,22.3680,2,0.20,2.5164,2015,October,10,4,7,11.25
5,6,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,48.8600,7,0.00,14.1694,2014,June,6,2,5,29.00
6,7,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,7.2800,4,0.00,1.9656,2014,June,6,2,5,27.00
7,8,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,907.1520,6,0.20,90.7152,2014,June,6,2,5,10.00
8,9,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,18.5040,3,0.20,5.7825,2014,June,6,2,5,31.25
9,10,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,114.9000,5,0.00,34.4700,2014,June,6,2,5,30.00


**2. Filtrar información (WHERE)**

Mostrar las ventas con una ganancia superior a 3,000.

  ***Objetivo:**
  Identificar las ventas con mayor rentabilidad.*

In [307]:
con.execute("""
SELECT
    "Order_ID",
    "Customer_Name",
    Category,
    Sales,
    Profit
FROM store
WHERE Profit > 3000
""").fetch_df()

,order_id,customer_name,category,sales,profit
0,CA-2015-145352,Christopher Martinez,Office Supplies,6354.950,3177.4750
1,CA-2017-127180,Tom Ashbrook,Technology,11199.968,3919.9888
2,CA-2014-116904,Sanjit Chand,Office Supplies,9449.950,4630.4755
3,CA-2017-166709,Hunter Lopez,Technology,10499.970,5039.9856
4,CA-2016-118689,Tamara Chand,Technology,17499.950,8399.9760
5,CA-2017-140151,Raymond Buch,Technology,13999.960,6719.9808
6,CA-2016-117121,Adrian Barton,Office Supplies,9892.740,4946.3700


**3. Agrupación y agregación (GROUP BY)**

Calcular las ventas totales por categoría.

  ***Objetivo:**
      Comparar el volumen de ventas entre las diferentes categorías de productos.*

In [308]:
con.execute("""
SELECT
    Category,
    SUM(Sales) AS Total_Sales
FROM store
GROUP BY Category
""").fetch_df()

,category,Total_Sales
0,Furniture,741999.7953
1,Office Supplies,719047.0320
2,Technology,836154.0330


**4. Ranking (ORDER BY)**

Obtener los cinco estados con mayores ventas.

  ***Objetivo:**
Identificar los estados con mayor volumen de ventas.*

In [309]:
con.execute("""
SELECT
    State,
    SUM(Sales) AS Total_Sales
FROM df
GROUP BY State
ORDER BY Total_Sales DESC
LIMIT 5;
""").fetch_df()

,state,Total_Sales
0,California,457687.6315
1,New York,310876.2710
2,Texas,170188.0458
3,Washington,138641.2700
4,Pennsylvania,116511.9140


**5. Consulta combinada (WHERE + GROUP BY + ORDER BY + cálculo)**

Calcular las ventas y ganancias por categoría considerando únicamente las ventas superiores a 100 dólares.

  ***Objetivo:**
Analizar la rentabilidad de cada categoría considerando únicamente ventas de mayor valor.*

In [310]:
con.execute("""
SELECT
    Category,
    SUM(Sales) AS Total_Sales,
    SUM(Profit) AS Total_Profit,
    ROUND(AVG(Profit),2) AS Average_Profit
FROM df
WHERE Sales > 100
GROUP BY Category
ORDER BY Total_Profit DESC;
""").fetch_df()

,category,Total_Sales,Total_Profit,Average_Profit
0,Technology,802689.8800,140006.4332,117.95
1,Office Supplies,587882.7070,97126.2641,77.08
2,Furniture,710789.0129,15919.3330,12.08


# Hallazgos y Conclusiones

Durante el análisis exploratorio del dataset Superstore Sales se identificaron los siguientes aspectos relevantes:

- El conjunto de datos presenta información completa sobre pedidos, ventas, ganancias, descuentos y categorías de productos.
- Se verificó la calidad de los datos mediante la revisión de valores nulos y registros duplicados.
- La categoría de mayor volumen de ventas fue Tecnología.
- Los 5 estados con mayor contribución al volumen de ventas son California, New York, Washington, Texas y Pensilvania.
- La tendencia de ventas de forma mensual va creciendo y tiene un proyección al incremento.
- Los gráficos interactivos facilitaron la identificación de tendencias y valores atípicos.
- Las consultas SQL con DuckDB permitieron obtener información resumida sobre ventas, ganancias y comportamiento por categorías y regiones de forma eficiente.

En conclusión, el análisis exploratorio permitió comprender el comportamiento comercial de la empresa y obtener información útil para apoyar la toma de decisiones relacionadas con ventas, rentabilidad y segmentación de clientes.